tm2/pkg/sdk/params/keeper.go

```diff
@@ -145,34 +153,85 @@ func (pk ParamsKeeper) SetBytes(ctx sdk.Context, key string, value []byte) {
 	pk.set(ctx, key, value)
 }
 
-func (pk ParamsKeeper) getIfExists(ctx sdk.Context, key string, ptr interface{}) {
+// GetParam gets a param value from the global param store.
+// Users generally should not cache anything and instead rely on the efficiency
+// of paramk.GetParams().
+func (pk ParamsKeeper) GetParams(ctx sdk.Context, moduleKey string, key string, target interface{}) (bool, error) {
+	if moduleKey != "" {
+		if pk.IsRegistered(moduleKey) {
+			key = moduleKey + ":" + key
+		} else {
+			return false, fmt.Errorf("params module key %q does not exisit", moduleKey)
+		}
+	}
+
 	stor := ctx.Store(pk.key)
-	bz := stor.Get([]byte(key))
+	vk := ValueStoreKey(key)
+	bz := stor.Get(vk)
 	if bz == nil {
-		return
+		return false, nil
 	}
-	err := amino.UnmarshalJSON(bz, ptr)
+
+	return true, amino.UnmarshalJSON(bz, target)
+}
+
+// SetParam sets a param value to the global param store.
+func (pk ParamsKeeper) SetParams(ctx sdk.Context, moduleKey string, key string, param interface{}) error {
+	if moduleKey != "" {
+		if pk.IsRegistered(moduleKey) {
+			key = moduleKey + ":" + key
+		} else {
+			return fmt.Errorf("parameter module key %q does not exist", moduleKey)
+		}
+	}
+
+	bz, err := amino.MarshalJSON(param)
```

Comment:
Rather than storing the whole thing under "key" as a single object {...}, 
every field in param should be broken up, and stored either as "module:_.key" or "module.key". 
This requires reflection, like reflect.ValueOf(param).Field(i).Tag and .Name, and whatever logic amino uses to derive the json key name.